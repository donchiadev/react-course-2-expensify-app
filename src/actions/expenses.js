import uuid from 'uuid'

// ADD_EXPENSE action generator
const addExpense = (
    { 
        description = '', 
        note = '', 
        amount = 0,
        createdAt = 0
    } = {}
) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
})

// REMOVE_EXPENSE action generator
const removeExpense = ({ id } = { }) => ({
    type: 'REMOVE_EXPENSE',
    id
})

// EDIT_EXPENSE action generator
const editExpense = (id, updates = { }) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export { addExpense, removeExpense, editExpense } 