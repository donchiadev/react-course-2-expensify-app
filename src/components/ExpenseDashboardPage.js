import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseSummary from './ExpensesSummary'

export const ExpenseDashboardPage = () => (
    <div>
        <ExpenseListFilters />
        <ExpenseSummary />
        <ExpenseList />
    </div>
)

export default ExpenseDashboardPage