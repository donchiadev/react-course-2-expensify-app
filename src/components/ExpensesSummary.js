import React from 'react'
import { connect } from "react-redux";
import numeral from 'numeral'
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total'

export const ExpenseSummary = (props) => (
    <div>
        <h3>
            {
                props.expensesCount === 1 ? (
                    "There is 1 visible expense."
                ) : (
                    `There are ${props.expensesCount} visible expenses.`
                )
            }
        </h3>
        <h3>Expenses Total is: {numeral(props.expensesTotal / 100).format('$0,0.00')}</h3>
    </div>
)

const mapStateToProps = (state) => {
    const {expenses, filters} = state
    const visibileExpenses = selectExpenses(expenses, filters)
    return {
        expensesCount: visibileExpenses.length,
        expensesTotal: selectExpensesTotal(visibileExpenses)
    }
}

export default connect(mapStateToProps)(ExpenseSummary)