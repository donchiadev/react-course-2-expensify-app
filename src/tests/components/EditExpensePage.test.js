import React from 'react'
import { shallow } from 'enzyme'
import { EditExpensePage } from '../../components/EditExpensePage'
import expenses from '../fixtures/expenses'

let history, editExpense, removeExpense, wrapper

beforeEach(() => {
    editExpense = jest.fn()
    removeExpense = jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(<EditExpensePage 
        expense={expenses[0]} 
        history={history} 
        editExpense={editExpense} 
        removeExpense={removeExpense} />)
})

test('should render EditExpensePage', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should handle editExpense', () => {
    const updates = { description: 'Laptop', note: 'Buyed a new laptop', amount: 80000 }
    wrapper.find('ExpenseForm').prop('onSubmit')(updates)
    expect(history.push).toHaveBeenCalledWith('/')
    expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, updates)
})

test('should removeExpense', () => {
    wrapper.find('button').at(0).prop('onClick')()
    expect(history.push).toHaveBeenCalledWith('/')
    expect(removeExpense).toHaveBeenLastCalledWith(expenses[0].id)
})

