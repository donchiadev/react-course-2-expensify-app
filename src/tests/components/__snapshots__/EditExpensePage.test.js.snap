// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <h1>
    Edit Expense
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 109500,
        "createdAt": 0,
        "description": "Rent",
        "id": "1",
        "note": "Just the rent bill",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
